### PR TITLE
Feature blockchain event subsciption

### PR DESCRIPTION
**Pull request by pawelkisielewicz**
Resolves https://github.com/golemfactory/concent/issues/899

Please check careful this pull request, the main part is to add callback in `finalize_deposit_claim`. It is not possible to test it local and without @kbeker I don't think we can test it with blockchain.  

There is also added test for function removing deposit_claim. But this function actually checks if object exists in database and calls other function, which is tested. I would remove this (my) test, but please comment

### CODE REVIEW

**Comment by dybi** [comment]
@pawelkisielewicz 
> There is also added test for function removing deposit_claim. But this function actually checks if object exists in database and calls other function, which is tested. I would remove this (my) test, but please comment

I do not understand. Could you please clarify what you mean?

File: concent_api/core/subtask_helpers.py [hunk 1]
```diff
@@ -23,6 +23,8 @@
 from core.models import PendingResponse
 from core.models import Subtask
 from core.payments import bankster
+from core.payments.bankster import remove_deposit_claim_if_transaction_successed_in_blockchain
```
Comment (Jakub89) [inline concent_api/core/subtask_helpers.py]: You can remove that import and use `bankster.remove_deposit_claim_if_transaction_successed_in_blockchain` in call below, or remove `bankster` import and import all used functions separately, but I don't know if that's possible.

**Comment by pawelkisielewicz** [comment]
There is function:
```python
def remove_deposit_claim_if_transaction_successed_in_blockchain(deposit_claim: DepositClaim) -> None:
    if DepositClaim.objects.filter(pk=deposit_claim.pk).exists():
        discard_claim(deposit_claim)
```
I wrote test to this function, but it is actually duplication of test to function  `def discard_claim`. I don't think test for `remove_deposit_claim_if_transaction_successed_in_blockchain` is necessary. What do you think?

File: concent_api/core/subtask_helpers.py [hunk 2]
```diff
@@ -277,7 +279,7 @@ def finalize_deposit_claim(
     ethereum_address: str,
     public_key: str,
 ) -> None:
-    deposit_claim = get_one_or_none(
+    deposit_claim: DepositClaim = get_one_or_none(  # type: ignore
```
Comment (dybi) [inline concent_api/core/subtask_helpers.py]: something's wrong here. `DepositClaim` and `# type: ignore` at the same time?

File: concent_api/core/subtask_helpers.py (hunk 1 above)
Comment (pawelkisielewicz) [reply concent_api/core/subtask_helpers.py]: done, import removed and now function is called:
`bankster.remove_deposit_claim_if_transaction_successed_in_blockchain`

File: concent_api/core/payments/bankster.py [hunk 3]
```diff
@@ -391,3 +391,8 @@ def discard_claim(deposit_claim: DepositClaim) -> bool:
             claim_removed = True
 
     return claim_removed
+
+
+def remove_deposit_claim_if_transaction_successed_in_blockchain(deposit_claim: DepositClaim) -> None:
+    if DepositClaim.objects.filter(pk=deposit_claim.pk).exists():
```
Comment (dybi) [inline concent_api/core/payments/bankster.py]: is there a possibility that this `deposit_claim` doesn't exist at this point?

Comment (dybi) [reply concent_api/core/payments/bankster.py]: BTW:
"successed" -> "succeeded"

File: concent_api/core/subtask_helpers.py (hunk 2 above)
Comment (pawelkisielewicz) [reply concent_api/core/subtask_helpers.py]: see `get_one_or_none` return type. It returns Optional[Subtask]. This function was created to be generic, for different objects, not only subtask. Signature was changed in the meantime (I don't think it was good change), but function still works for different objects. 

In this case `DepositClaim` is passed as Model, so its instance is return type. I added `: DepositClaim` to use real/correct type, but ignoring type was also necessary. 

BTW.  typing in `get_one_or_none` should be fixed. What do you think?

File: concent_api/core/payments/bankster.py (hunk 3 above)
Comment (pawelkisielewicz) [reply concent_api/core/payments/bankster.py]: good idea :+1: . It is checked by function calling, so there is no need to check it again

**Comment by pawelkisielewicz** [comment]
@dybi @rwrzesien can you please check if part with creating callback with passing function using lambda is OK. This is the most important part here!

File: concent_api/core/payments/bankster.py [hunk 4]
```diff
@@ -145,7 +145,7 @@ def claim_deposit(
             # If the total of existing claims and the current claim is greater or equal to the current deposit,
             # we can't add a new claim.
             if provider_deposit <= sum_of_existing_provider_claims['sum_of_existing_claims'] + settings.ADDITIONAL_VERIFICATION_COST:
-                return (requestor_deposit, None)
+                return (None, None)
```
Comment (cameel) [inline concent_api/core/payments/bankster.py]: This looks like a bugfix for something that was broken before this pull request. It should be a separate commit.

File: concent_api/core/payments/bankster.py [hunk 5]
```diff
@@ -239,6 +239,7 @@ def finalize_payment(deposit_claim: DepositClaim) -> Optional[str]:
 
         # Bankster puts transaction ID in DepositClaim.tx_hash.
         deposit_claim.tx_hash = encode_hex(ethereum_transaction.hash)
+        deposit_claim.full_clean()
```
Comment (cameel) [inline concent_api/core/payments/bankster.py]: Also a bugfix. Separate commit.

File: concent_api/core/subtask_helpers.py (hunk 2 above)
Comment (cameel) [reply concent_api/core/subtask_helpers.py]: `get_one_or_none` definitely should get fixed. It was meant to work with any model.

File: concent_api/core/subtask_helpers.py [hunk 6]
```diff
@@ -311,7 +313,12 @@ def finalize_deposit_claim(
     )
 
     if deposit_claim:
```
Comment (cameel) [inline concent_api/core/subtask_helpers.py]: It's a model, not a boolean. It's not a change from your pull request but the fix is trivial so please change it to `if deposit_claim is not None:`.

File: concent_api/core/subtask_helpers.py [hunk 7]
```diff
@@ -311,7 +313,12 @@ def finalize_deposit_claim(
...
-        bankster.finalize_payment(deposit_claim)
+        tx_hash = bankster.finalize_payment(deposit_claim)
+        payment_interface = PaymentInterface()  # new object is NOT created, it is singleton and its instance is returned
+        payment_interface.on_transaction_confirmed(   # type: ignore  # pylint: disable=no-member
+            tx_hash=tx_hash,
+            cb=lambda: discard_claim(deposit_claim)
+        )
```
Comment (cameel) [inline concent_api/core/subtask_helpers.py]: Please make sure that `discard_claim()` does not fail if two processes try to do it at the same time (i.e. first gets a lock, seconds waits on the lock, first runs `delete()` and releases the lock, second runs `delete()` on the same, already deleted object). I think it won't fail on the SQL level (the `WHERE` clause in the second `DELETE` will simply match no rows) but I'm not sure how Django will behave.

File: concent_api/core/payments/bankster.py (hunk 4 above)
Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: I think it is now seen because branch is not rebased. It must `return (requestor_deposit, None)`

Comment (cameel) [reply concent_api/core/payments/bankster.py]: But why? It's supposed to return a claim instance. `requestor_deposit` is a number.

File: concent_api/core/subtask_helpers.py (hunk 6 above)
Comment (pawelkisielewicz) [reply concent_api/core/subtask_helpers.py]: done
